### PR TITLE
[Not Ready] Adds ability to push defaults to client/ client builder to adds default client nodes and a client builder

### PR DESCRIPTION
**Pull request by QuestofIranon**

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -20,20 +20,52 @@ use crate::{
...
 };
 
+pub struct ClientBuilder {
+    address: String,
+    node: Option<AccountId>,
+}
+
 pub struct Client {
-    pub(crate) inner: Arc<grpc::Client>,
+    raw_client: Arc<grpc::Client>,
```
Comment (mehcode) [inline src/client.rs]: Why the name change from `inner` to `raw_client` ?

File: src/client.rs [hunk 2]
```diff
@@ -20,20 +20,52 @@ use crate::{
...
+        }
+
+        Ok(client)
+    }
+
 }
 
 impl Client {
-    pub fn new(address: impl AsRef<str>) -> Result<Self, Error> {
+    pub fn builder(address: impl AsRef<str>) -> ClientBuilder {
```
Comment (mehcode) [inline src/client.rs]: `ClientBuilder`s are meant to last for a very short time. Storing `&'a str` inside should be fine.

File: src/client.rs [hunk 3]
```diff
@@ -20,20 +20,52 @@ use crate::{
...
+        ClientBuilder{
+            address: address.as_ref().into(),
+            node: None
+        }
+    }
+
+    pub(crate) fn new(address: &impl AsRef<str>) -> Result<Self, Error> {
```
Comment (mehcode) [inline src/client.rs]: `&impl AsRef<str>` is redundant.. I think.

File: src/client.rs [hunk 4]
```diff
@@ -20,20 +20,52 @@ use crate::{
...
+}
+
 pub struct Client {
-    pub(crate) inner: Arc<grpc::Client>,
+    raw_client: Arc<grpc::Client>,
+    node: Option<AccountId>,
+}
+
+impl ClientBuilder {
+    pub fn node(&mut self, node: AccountId) -> &mut Self {
```
Comment (mehcode) [inline src/client.rs]: This should probably be a consuming builder so you can move memory easily at the end. E.g.. each method should take `mut self` and return `Self`.

File: src/client.rs [hunk 5]
```diff
@@ -45,15 +77,36 @@ impl Client {
...
+        self.raw_client.clone()
     }
 
     /// Create a new account. After the account is created, the AccountID for it is in the
     /// receipt, or can be retrieved with a GetByKey query, or by asking for a Record of the
     /// transaction to be created, and retrieving that.
     #[inline]
     pub fn create_account(&self) -> Transaction<TransactionCryptoCreate> {
-        TransactionCryptoCreate::new(self)
+        let mut tx = TransactionCryptoCreate::new(self);
```
Comment (mehcode) [inline src/client.rs]: After thinking on this more.. this makes way more sense to do in `Transaction::new`. That gets passed the client. Just make the fields `pub(crate)`.

**Review by mehcode — CHANGES_REQUESTED** [review]

File: src/client.rs (hunk 3 above)
Comment (QuestofIranon) [reply src/client.rs]: This needs to be directly usable from the bridge though, right?

Comment (mehcode) [reply src/client.rs]: But a reference to a `AsRef` seems odd

Comment (QuestofIranon) [reply src/client.rs]: ohh I see what you mean. I didn't notice the unneeded `&`

File: src/client.rs (hunk 1 above)
Comment (QuestofIranon) [reply src/client.rs]: Left over from a previous version of the change.

**Comment by QuestofIranon** [comment]
Don't merge this yet, I'm going to add the default operator to this branch/ pr

**Comment by QuestofIranon** [comment]
Yeah don't merge this yet, need to clean up the builder after rebasing and I'll add the default operator with this PR as well.

File: src/transaction/transaction.rs [hunk 6]
```diff
@@ -86,9 +88,10 @@ impl<T: 'static> Transaction<T, TransactionBuilder<T>> {
         self
     }
 
-    pub fn operator(&mut self, id: AccountId) -> &mut Self {
+    pub fn operator(&mut self, id: AccountId, secret: SecretKey) -> &mut Self {
```
Comment (mehcode) [inline src/transaction/transaction.rs:88]: Operator secret should _not_ be provided via normal builder methods. If operator is changed however we should invalidate the stored secret.

File: src/transaction/transaction.rs [hunk 7]
```diff
@@ -156,6 +161,12 @@ impl<T: 'static> Transaction<T, TransactionBuilder<T>> {
             }
         }
 
+        if let Some(state) = self.as_builder() {
```
Comment (mehcode) [inline src/transaction/transaction.rs]: Just as a heads up. The operator must be the **first** signature. I would insert it to the top of the vec right before send.

And there is some code in sign that checks for the 2nd signature and does some special stuff. If we have an operator signature then it needs to check for the first signature.

File: src/transaction/transaction.rs [hunk 8]
```diff
@@ -58,13 +59,19 @@ impl<T: 'static> Transaction<T, TransactionBuilder<T>> {
     where
         T: Object + ToProto<proto::Transaction::TransactionBody_oneof_data> + 'static,
     {
+        let id = match client.operator {
```
Comment (mehcode) [inline src/transaction/transaction.rs]: `client.operator.map(TransactionId::new)`

File: src/transaction/transaction.rs [hunk 9]
```diff
@@ -58,13 +59,19 @@ impl<T: 'static> Transaction<T, TransactionBuilder<T>> {
...
+            Some(operator_id) => Some(TransactionId::new(operator_id)),
+            None => None
+        };
+
         Self {
             crypto_service: client.crypto.clone(),
             file_service: client.file.clone(),
             contract_service: client.contract.clone(),
+            secret: client.operator_secret.clone(),
```
Comment (mehcode) [inline src/transaction/transaction.rs:66]: This should be behind an Arc

File: src/client.rs [hunk 10]
```diff
@@ -13,35 +13,81 @@ use crate::{
...
+    }
+
+    pub fn build(self) -> Result<Client, Error> {
+
+        let mut client = Client::new(&self.address)?;
+
+        if let Some(node) = self.node {
+            client.set_node(node);
+        }
+
```
Comment (mehcode) [inline src/client.rs:66]: Forgot to call `set_operator`

File: src/client.rs [hunk 11]
```diff
@@ -13,35 +13,81 @@ use crate::{
...
+    pub(crate) fn new(address: impl AsRef<str>) -> Result<Self, Error> {
         let address = address.as_ref();
         let (host, port) = address.split(':').next_tuple().ok_or_else(|| {
             format_err!("failed to parse 'host:port' from address: {:?}", address)
         })?;
 
         let port = port.parse()?;
 
-        let inner = Arc::new(grpc::Client::new_plain(
+        let raw_client = Arc::new(grpc::Client::new_plain(
```
Comment (mehcode) [inline src/client.rs]: `raw_client` -> `inner` or `client`.

**Pull request merged**